Fixed Ollama Structured Response not working  #10616

imdigitalashish opened the pull request:
## Title

feat(ollama): Allow direct Pydantic model usage in response_format for ollama_chat

## Relevant issues

Fixes #10616

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] have added a screenshot of my new test passing locally
- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem


## Type

🐛 Bug Fix


## Changes

The key changes are:

1.  **Bug Resolution**: This resolves the underlying issue that led to `KeyError: 'schema'` when users attempted to use Pydantic models without the precise nested structure, as reported in #10616. The PR ensures the schema is correctly processed by `litellm` before being passed to the Ollama API.

These changes enhance the developer experience for using structured outputs with Ollama via `litellm` and address the incompatibility noted in the referenced issue.

<img width="643" alt="image" src="https://github.com/user-attachments/assets/898e8c04-6997-4745-8788-78c0d644e9d6" />

imdigitalashish committed:
Fixed Json.dumps in JSON Schema Validation Error

imdigitalashish committed:
Added Response Schema to Ollama chat for structured response

imdigitalashish committed:
Added Test cases

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/BerriAI/litellm?pullRequest=10617) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/BerriAI/litellm?pullRequest=10617) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/BerriAI/litellm?pullRequest=10617) it.</sub>

krrishdholakia commented on tests/certifi-2025.4.26-py3-none-any.whl:
please fix your gitignore / remove these .whl files

imdigitalashish replied:
Done

krrishdholakia commented on litellm/llms/ollama_chat.py:
```diff
@@ -141,6 +142,14 @@ def map_openai_params(
         model: str,
         drop_params: bool,
     ) -> dict:
+        if "response_format" in non_default_params:
```
this should not be needed. the response_Format conversion to a json is handled in get_optional_params already before this step - https://github.com/BerriAI/litellm/blob/d680feb5dc9b824e353e9dff8df252030a1ddbb6/litellm/utils.py#L2795

imdigitalashish replied:
This is fixed as well

imdigitalashish committed:
refactor(ollama): remove redundant response_format check

The response_format parameter conversion is already handled in utils.py's
get_optional_params function, making the duplicate check in ollama_chat.py
unnecessary. This change removes the redundant code while maintaining the
same functionality.

yelonek commented:
Hi @imdigitalashish  and @krrishdholakia, the lack of structured output in ollama via litellm drove me crazy far enough that I looked into it today.

I see you check for type `json_object` and `json_schema`, but what is sent as a parameter to `response_format` is the result of `model_json_schema()` and that is a dict where type is set to "object", not "json_object".

![image](https://github.com/user-attachments/assets/94952de8-8577-4e91-b8af-1173cb10cd83)

I just added additional if and then it worked as expected:
```
if param == "response_format" and value["type"] == "object":
                optional_params["format"] = value
```
I am using pydantic '2.11.4' and litellm '1.69.1'

Let me know what you think!

krrishdholakia commented:
Hey @yelonek is this an alternative to @imdigitalashish's PR? 

if so - can you file it separately with a unit test in `tests/litellm`?

yelonek commented:
> Hey @yelonek is this an alternative to @imdigitalashish's PR?
> 
> if so - can you file it separately with a unit test in `tests/litellm`?

I think it is complementary.
First I need to figure out how to set up the dev environment for litellm. So far I figured out, that Windows will not cut it.  ;)

krrishdholakia commented:
What's the issue you see? @yelonek
